pkg/services/pull_request/bitbucket_server.go

```diff
@@ -0,0 +1,97 @@
+package pull_request
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	bitbucketv1 "github.com/gfleury/go-bitbucket-v1"
+)
+
+type BitbucketService struct {
+	client         *bitbucketv1.APIClient
+	projectKey     string
+	repositorySlug string
+	branchMatch    *regexp.Regexp
+	// Not supported for PRs by Bitbucket Server
+	// labels         []string
+}
+
+var _ PullRequestService = (*BitbucketService)(nil)
+
+func NewBitbucketServiceBasicAuth(ctx context.Context, username, password, url, projectKey, repositorySlug string, branchMatch *string) (PullRequestService, error) {
+	bitbucketConfig := bitbucketv1.NewConfiguration(url)
+	// Avoid the XSRF check
+	bitbucketConfig.AddDefaultHeader("x-atlassian-token", "no-check")
+	bitbucketConfig.AddDefaultHeader("x-requested-with", "XMLHttpRequest")
+
+	ctx = context.WithValue(ctx, bitbucketv1.ContextBasicAuth, bitbucketv1.BasicAuth{
+		UserName: username,
+		Password: password,
+	})
+	return newBitbucketService(ctx, bitbucketConfig, projectKey, repositorySlug, branchMatch)
+}
+
+func NewBitbucketServiceNoAuth(ctx context.Context, url, projectKey, repositorySlug string, branchMatch *string) (PullRequestService, error) {
+	return newBitbucketService(ctx, bitbucketv1.NewConfiguration(url), projectKey, repositorySlug, branchMatch)
+}
+
+func newBitbucketService(ctx context.Context, bitbucketConfig *bitbucketv1.Configuration, projectKey, repositorySlug string, branchMatch *string) (PullRequestService, error) {
+	if !strings.HasSuffix(bitbucketConfig.BasePath, "/rest") {
+		bitbucketConfig.BasePath = bitbucketConfig.BasePath + "/rest"
```

Comment:
The library is more-or-less auto-generated and follows this example for all methods: `a.client.cfg.BasePath + "/api/1.0/projects/{projectKey}/repos"`
I believe they used the WADL provided by Bitbucket which defines the base path like this `base="http://example.com/rest/"`. This basically means that `/rest` is mandatory. Initially, I also wrote it explicitly (it's even in the example provided), but decided to handle it in code because it basically always has to be there. Ideally, it would've been part of the library.

I don't mind removing this and adding some extra docs for this, but do note that it always needs to be there and "regular" scm users wouldn't know this detail without the example in the docs.

Reply:
Sorry for the response delay, refactored it a bit and added support for `/rest/`